---
title: "Analyzing fixed differences between species"
author: "Miwa Wenzel"
date: 5/2/2020
output: github_document
---

# Set up
## Input files
### ***CHANGE THESE IF YOU WISH***
#### Input aligned sequences as two separate fasta files.

```{r input}

###melanogaster and simulans
input_fastaA = "~/github/Sequence_Analysis/input_data/Dmel_aligned.fasta" # ***CHANGE*** to your first sequence of interest
input_fastaB = "~/github/Sequence_Analysis/input_data/Dsim_aligned.fasta" # ***CHANGE*** to your second sequence of interest

```

#### Set up labels and stuff for the output graphs
```{r labels}

subtitle_label = "bam | D. mel and D. sim" # ***CHANGE*** to reflect your sequences

######### SET FIGURE SIZES ##########

histwidth = 5 # ***CHANGE*** to set the width of the divergence output histogram in inches
histheight = 5 # ***CHANGE*** to set the width of the divergence output histogram in inches

plotwidth = 14 # ***CHANGE*** to set the width of the output plot in inches
plotheight = 3 # ***CHANGE*** to set the width of the output plot in inches

######### SET UP REGIONS TO ANNOTATE ON THE PLOT #########

rect_df = data.frame(xmin = c(151, 13, 402, 201), # ***CHANGE*** if want to highlight different regions for your gene
                     xmax = c(350, 33, 434, 250), # ***CHANGE*** to correspond to ends of the regions that you put in the above "xmin" section (e.g. the first region is from amino acid position 13 to 33)
                     ymin = c(-Inf, -Inf, -Inf, -Inf),
                     ymax = c(Inf, Inf, Inf, Inf),
                     Region = factor(c("Bgcn/Csn4 binding", "CBM", "PEST domain", "Ubiquitin interaction"))) # ***CHANGE*** to reflect the names of your regions

hypomorph = data.frame(loc = 255, hypomorph = "L255F") # ***CHANGE*** if you want to highlight any position with a vertical line, e.g. like it currently points to the hypomorph position; I can't get this to show up in the legend however.

```

## Functions
### ***DON'T CHANGE THESE OR ANYTHING ELSE BELOW... unless you wanna fix my code :)***
##### This is a set of functions to get sequence data into a few different usable forms. Not all may be relevant for this code, but I maintain them together for ease. 

```{r sequence_functions}
############# IMPORT FASTA FILES AND CREATE DATAFRAME OF THE SEQUENCES ###################

library(seqinr) #load seqinr library in order to use the read.fasta function below

get_sequences = function(fasta) { #import sequences from fasta format
  sequence_list = read.fasta(fasta, seqtype = "DNA", seqonly = TRUE)
  sequence_list2 = gsub(" ", "", sequence_list) #remove any spaces in the fasta file if they are present to make sequences compatible with downstream functions
  sequence_names = c() #set empty vector to create sequence names to identify each sequence in the fasta file

  for (i in 1:length(sequence_list)) { #for each sequence in the fasta file, give it an identity of "sequence1, sequence2, etc." Then compile these such that we have a dataframe where the first column is the sequence names and the second column is the corresponding sequence itself)
    sequence_name = {}
    sequence_name = paste("sequence", i, sep = "")
    sequence_names = c(sequence_names, sequence_name)
  }

  sequence = mapply(assign, sequence_names, sequence_list2)
  as.data.frame(sequence)

}



############# CREATE A DATAFRAME OF NUCLEOTIDES FOR THE SEQUENCES ##################

library(stringr) #load library to use str_split_fixed function below

get_nucleotides = function(Sequences) { #Create a dataframe such that each nucleotide is in its own column. Thus we get a very large matrix where the rows are the different sequences and the columns are the nucleotides in the sequences.

  nucleotides = str_split_fixed(Sequences$sequence, "", str_length(Sequences[1,1]))
}

############# CREATE A DATAFRAME OF CODONS FOR THE SEQUENCES ##############

get_codons = function(Sequences) { #Now instead of splitting up the sequences into individual nucleotides for a dataframe, we want to split it up into codons
  Codons = data.frame(matrix(nrow = nrow(Sequences), ncol = str_length(Sequences[1,1])/3))

  for (b in 1:nrow(Sequences)) {
    codon_vector = c()

    for (i in seq(1, str_length(Sequences[b,1]), 3)) { #break up the sequences into codons
      seq_codon = substr(Sequences[b,1], i, i + 2)
      codon_vector = c(codon_vector, seq_codon)
    }
    Codons[b,] = codon_vector
  }
  codons = Codons
}

############# CREATE A DATAFRAME OF AMINO ACIDS FOR THE SEQUENCES #############

codon_to_aa = function(dna) {

 switch(dna, 
        "ATA" = "I", "ATT" = "I", "ATC" = "I",
        "CTT" = "L", "CTC" = "L", "CTA" = "L", "CTG" = "L", "TTA" = "L", "TTG" = "L",
        "GTT" = "V", "GTC" = "V", "GTA" = "V", "GTG" = "V",
        "TTT" = "F", "TTC" = "F",
        "ATG" = "M",
        "TGT" = "C", "TGC" = "C",
        "GCT" = "A", "GCC" = "A", "GCA" = "A", "GCG" = "A",
        "GGT" = "G", "GGC" = "G", "GGA" = "G", "GGG" = "G",
        "CCT" = "P", "CCC" = "P", "CCA" = "P", "CCG" = "P",
        "ACT" = "T", "ACC" = "T", "ACA" = "T", "ACG" = "T",
        "TCT" = "S", "TCC" = "S", "TCA" = "S", "TCG" = "S", "AGT" = "S", "AGC" = "S",
        "TAT" = "Y", "TAC" = "Y",
        "TGG" = "W",
        "CAA" = "Q", "CAG" = "Q",
        "AAT" = "N", "AAC" = "N",
        "CAT" = "H", "CAC" = "H",
        "GAA" = "E", "GAG" = "E",
        "GAT" = "D", "GAC" = "D",
        "AAA" = "K", "AAG" = "K",
        "CGT" = "R", "CGC" = "R", "CGA" = "R", "CGG" = "R", "AGA" = "R", "AGG" = "R",
        "TAA" = "Stop", "TAG" = "Stop", "TGA" = "Stop", "NNN" = "NA",
        "CNN" = "NA", "TNN" = "NA", "GNN" = "NA", "ANN" = "NA", 
        "CCN" = "NA", "TTN" = "NA", "GGN" = "NA", "AAN" = "NA", 
        "NNC" = "NA", "NNT" = "NA", "NNG" = "NA", "NNA" = "NA", 
        "NCC" = "NA", "NTT" = "NA", "NGG" = "NA", "NAA" = "NA", 
        "CTN" = "NA", "TCN" = "NA", "NCT" = "NA", "NTC" = "NA", 
        "CAN" = "NA", "ACN" = "NA", "NCA" = "NA", "NAC" = "NA", 
        "CGN" = "NA", "GCN" = "NA", "NCG" = "NA", "NGC" = "NA", 
        "ATN" = "NA", "TAN" = "NA", "NAT" = "NA", "NTA" = "NA", 
        "AGN" = "NA", "GAN" = "NA", "NAG" = "NA", "NGA" = "NA", 
        "GTN" = "NA", "TGN" = "NA", "NGT" = "NA", "NTG" = "NA") 
}

get_AAs = function(Codon_df) {
  AAs = apply(Codon_df, c(1,2), codon_to_aa)

}

```

### This is the set up for getting the Miyata scores for the polymorphisms
##### Use of hash function taken from here: https://blog.ephorie.de/hash-me-if-you-can

```{r Miyata_function, results=FALSE}

assign_hash = Vectorize(assign, vectorize.args = c("x", "value"))

get_hash = Vectorize(get, vectorize.args = c("x"))

exists_hash = Vectorize(exists, vectorize.args = c("x"))

hash = new.env(hash = TRUE, parent = emptyenv(), size = 418L)

key = c("C P", "C A", "C G", "C S", "C T", "C Q", "C E", "C N", "C D", "C H", "C K", "C R", "C V", "C L", "C I", "C M", "C F", "C Y", "C W", 
        "P A", "P G", "P S", "P T", "P Q", "P E", "P N", "P D", "P H", "P K", "P R", "P V", "P L", "P I", "P M", "P F", "P Y", "P W", 
        "A G", "A S", "A T", "A Q", "A E", "A N", "A D", "A H", "A K", "A R", "A V", "A L", "A I", "A M", "A F", "A Y", "A W",
        "G S", "G T", "G Q", "G E", "G N", "G D", "G H", "G K", "G R", "G V", "G L", "G I", "G M", "G F", "G Y", "G W", 
        "S T", "S Q", "S E", "S N", "S D", "S H", "S K", "S R", "S V", "S L", "S I", "S M", "S F", "S Y", "S W", 
        "T Q", "T E", "T N", "T D", "T H", "T K", "T R", "T V", "T L", "T I", "T M", "T F", "T Y", "T W",
        "Q E", "Q N", "Q D", "Q H", "Q K", "Q R", "Q V", "Q L", "Q I", "Q M", "Q F", "Q Y", "Q W",
        "E N", "E D", "E H", "E K", "E R", "E V", "E L", "E I", "E M", "E F", "E Y", "E W",
        "N D", "N H", "N K", "N R", "N V", "N L", "N I", "N M", "N F", "N Y", "N W",
        "D H", "D K", "D R", "D V", "D L", "D I", "D M", "D F", "D Y", "D W",
        "H K", "H R", "H V", "H L", "H I", "H M", "H F", "H Y", "H W",
        "K R", "K V", "K L", "K I", "K M", "K F", "K Y", "K W",
        "R V", "R L", "R I", "R M", "R F", "R Y", "R W",
        "V L", "V I", "V M", "V F", "V Y", "V W",
        "L I", "L M", "L F", "L Y", "L W",
        "I M", "I F", "I Y", "I W",
        "M F", "M Y", "M W",
        "F Y", "F W",
        "Y W",

        "C NA", "P NA", "A NA", "G NA", "S NA", "T NA", "Q NA", "E NA", "N NA", "D NA", "H NA", "K NA", "R NA", "V NA", "L NA", "I NA", "M NA", "F NA", "Y NA",

        "P C", "A C", "G C", "S C", "T C", "Q C", "E C", "N C", "D C", "H C", "K C", "R C", "V C", "L C", "I C", "M C", "F C", "Y C", "W C", 
        "A P ", "G P", "S P", "T P", "Q P", "E P", "N P", "D P", "H P", "K P", "R P", "V P", "L P", "I P", "M P", "F P", "Y P", "W P", 
        "G A", "S A", "T A", "Q A", "E A", "N A", "D A", "H A", "K A", "R A", "V A", "L A", "I A", "M A", "F A", "Y A", "W A",
        "S G", "T G", "Q G", "E G", "N G", "D G", "H G", "K G", "R G", "V G", "L G", "I G", "M G", "F G", "Y G", "W G", 
        "T S", "Q S", "E S", "N S", "D S", "H S", "K S", "R S", "V S", "L S", "I S", "M S", "F S", "Y S", "W S", 
        "Q T", "E T", "N T", "D T", "H T", "K T", "R T", "V T", "L T", "I T", "M T", "F T", "Y T", "W T",
        "E Q", "N Q", "D Q", "H Q", "K Q", "R Q", "V Q", "L Q", "I Q", "M Q", "F Q", "Y Q", "W Q",
        "N E", "D E", "H E", "K E", "R E", "V E", "L E", "I E", "M E", "F E", "Y E", "W E",
        "D N", "H N", "K N", "R N", "V N", "L N", "I N", "M N", "F N", "Y N", "W N",
        "H D", "K D", "R D", "V D", "L D", "I D", "M D", "F D", "Y D", "W D",
        "K H", "R H", "V H", "L H", "I H", "M H", "F H", "Y H", "W H",
        "R K", "V K", "L K", "I K", "M K", "F K", "Y K", "W K",
        "V R", "L R", "I R", "M R", "F R", "Y R", "W R",
        "L V", "I V", "M V", "F V", "Y V", "W V",
        "I L", "M L", "F L", "Y L", "W L",
        "M I", "F I", "Y I", "W I",
        "F M", "Y M", "W M",
        "Y F", "W F",
        "W Y",

        "NA C", "NA P", "NA A", "NA G", "NA S", "NA T", "NA Q", "NA E", "NA N", "NA D", "NA H", "NA K", "NA R", "NA V", "NA L", "NA I", "NA M", "NA F", "NA Y"
        )



value = c(1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,

          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          
          1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,
          
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
          )

assign_hash(key, value, hash)

```

### Run the above functions to get sequence data in usable format
```{r sequences}

SequencesA = get_sequences(input_fastaA) #apply function in import sequences from fasta files
SequencesB = get_sequences(input_fastaB) 

NucleotidesA = get_nucleotides(SequencesA) #apply function to get the sequences into a usable dataframe format
NucleotidesB = get_nucleotides(SequencesB)

CodonsA = get_codons(SequencesA) #apply function to get sequences into codon format so we can subsequently translate to amino acids
CodonsB = get_codons(SequencesB)

AAsA = get_AAs(CodonsA) #apply function to get amino acids of the sequences
AAsB = get_AAs(CodonsB)

AA_aligned = rbind(AAsA, AAsB) #put the two translated sequences into the same dataframe
AA_aligned_df = as.data.frame(AA_aligned)

#Because the alignment was done with multiple species, there may be gaps in both of the chosen sequences used here and the for loop below gets rid of these. Don't run this for loop if you want all the sequences from the original alignent to be compared amongst the different graphs on the same x-axis (amion aicd) scale. 
remove_col_vector = c()
for (z in 1:ncol(AA_aligned_df)) { 
  if (as.character(AA_aligned_df[1, z]) == "NULL" && as.character(AA_aligned_df[2,z]) == "NULL") {
    remove_col = z
    remove_col_vector = c(remove_col_vector, z)
  }
}

AA_aligned_df_NoNULL = AA_aligned_df
AA_aligned_df_NoNULL[,c(remove_col_vector)] = NULL
```
# Analyses
## Do the Miyata analysis
```{r Miyata}

Miyata_vector = c()

for (i in 1:ncol(AA_aligned_df_NoNULL)) { #for every site in the amino acid alignment...
  
  if (as.character(AA_aligned_df_NoNULL[1, i]) == as.character(AA_aligned_df_NoNULL[2, i])) { #if the amino acids are the same...
#     print("Amino acids are conserved")
    Miyata = 0 #then the site is conserved and the Miyata score is 0
  } else if (AA_aligned_df_NoNULL[1,i] == "NULL" && AA_aligned_df_NoNULL[2,i] != "NULL") { #but if there is an amino acid in sequenceA but a gap in sequenceB...
    Miyata = 0 #then the Miyata score is also 0
  } else if (AA_aligned_df_NoNULL[2,i] == "NULL" && AA_aligned_df_NoNULL[1,i] != "NULL") { #and vice versa
    Miyata = 0
  } else { #otherwise, the amino acids are different...
        pairedAAs = paste(AA_aligned_df_NoNULL[1, i], AA_aligned_df_NoNULL[2,i])
#       print(pairedAAs)
    
        Miyata = hash[[pairedAAs]] #so get the Miyata scores of the amino acid pair
#       print(Miyata)
    }
  Miyata_vector = c(Miyata_vector, Miyata) #store the Miyata scores in a vector so we have them for every site
}
#  print(Miyata_vector)

#Miyata_vector_nt = rep.col(Miyata_vector, 3)
#Miyata_vector_ntgraph = as.vector(t(Miyata_vector_nt))

```

### A simple analysis of the distribution of Miyata scores of the nonsynonymous polymorphisms
```{r histogram, fig.height = histheight, fig.width = histwidth}

Miyata_vector_NoZero = Miyata_vector[!Miyata_vector %in% 0.00] #Get the Miyata scores for only sites that are nonsynonymous polymorphisms

############# MAKE THE HISTOGRAMS #############

library(ggplot2)
library(dplyr)

#A histogram of Miyata scores for every site in the alignment, i.e. includes lots of 0s for the conserved sites
#Not that useful, so it is commented out currently

#ggplot() +
#  aes(Miyata_vector) +
#  geom_histogram(binwidth = 0.25) +
#  labs(title = "Miyata scores of population", subtitle = subtitle_label, x = "Miyata score") +
#  scale_x_continuous(breaks = seq(0, 5.25, by = 1))

#A histogram of the set of Miyata scores of the divergences.

ggplot() +
  aes(Miyata_vector_NoZero) +
  geom_histogram(binwidth = 0.25) +
  labs(title = "Miyata scores of divergences", subtitle = subtitle_label, x = "Miyata score") +
  scale_x_continuous(breaks = seq(0, 5.25, by = 1)) #don't know why the x-scale isn't going to 5.25 here :/

```

## Make the final graph
##### Set up the background data. This is set to highlight documented functional domains of our gene bam. You can change this to reflect the known regions of your gene and/or not run it. ***I HAVE MOVED THIS TO THE TOP OF THE DOCUMENT SO IT IS EASY TO ACCESS AND CHANGE. Thus this code chunk is defaulted to NOT be run.***

```{r graph_background, eval=FALSE}

#Below is data used for amino acid regions. It is a dataframe of important regions to highlight; idk why, but order the regions in this way makes the legend be in the desired order

rect_df = data.frame(xmin = c(13, 151, 402, 201), #dataframe of important regions to highlight
                     xmax = c(33, 350, 434, 250),
                     ymin = c(-Inf, -Inf, -Inf, -Inf),
                     ymax = c(Inf, Inf, Inf, Inf),
                     Region = factor(c("bgcn/csn4 binding", "PEST domain", "Ubiquitin interaction")))

#Below is the data used for nucleotide positions
#rect_df = data.frame(xmin = c(451, 36, 1204, 601), 
#                     xmax = c(1050, 99, 1302, 750),
#                     ymin = c(-Inf, -Inf, -Inf, -Inf),
#                     ymax = c(Inf, Inf, Inf, Inf),
#                     Region = factor(c("CBM", "Bgcn/Csn4 binding", "PEST domain", "Ubiquitin interaction")))

hypomorph = data.frame(loc = 763, hypomorph = "L255F") #dataframe of hypomorph position

```

#### Set up the different types of plots to prepare for whether or not we want to include annotations in our final plot 
```{r plot_setup}

######## GET DATA INTO GRAPHABLE FORMAT ########
aminoacid_pos = c(1:ncol(AA_aligned_df_NoNULL))

divergence_Miyata_graph = rbind(aminoacid_pos, Miyata_vector)
divergence_Miyata_graph = as.data.frame(t(divergence_Miyata_graph))

graph_data_divergences_only = divergence_Miyata_graph[!(divergence_Miyata_graph$Miyata_vector==0.00),]


######## SET UP THE DIFFERENT PLOTS ########
library(ggplot2)

Hypomorph_SNP = ggplot() +     
            geom_vline(data = hypomorph, 
            aes(xintercept = loc), 
              color = "purple4") 

FxnlRegions = ggplot() +
            geom_rect(data = rect_df,
            aes(xmin = xmin, 
                xmax = xmax, 
                ymin = ymin,
                ymax = ymax,
                x = NULL,
                y = NULL,
                fill = Region,
                color = NULL),
            alpha = 0.3) +

  scale_fill_manual(labels = c("Bgcn/Csn4 binding", "CBM", "PEST domain", "Ubiquitin interaction"),
    values = c("steelblue1", "#CC79A7", "#009E73", "#F0E442"))

Divergences = ggplot() + 
    
  geom_point(data = graph_data_divergences_only, size = 3, shape = 2) +
  aes(x = aminoacid_pos,
      y = Miyata_vector) +
  labs(x = "Amino acid position",
       y = "Miyata score",
       title = "Miyata scores of fixed differences",
       subtitle = subtitle_label) +
    scale_x_continuous(limits = c(0, ncol(AA_aligned_df_NoNULL)))

```

### The final plot of divergences!
```{r final_plot, fig.height = plotheight, fig.width = plotwidth}

library(ggplot2)

if (exists("rect_df") == FALSE && exists("hypomorph") == FALSE) {
  print("Regions and a hypomorph SNP are NOT present to be annotated")
  final_plot = Divergences

} else if (exists("rect_df") == TRUE && exists("hypomorph") == FALSE) {
  print("Regions are present to be annotated")

  final_plot = FxnlRegions + 

  geom_point(data = graph_data_divergences_only, size = 3, shape = 2) +
  aes(x = aminoacid_pos,
      y = Miyata_vector) +
  labs(x = "Amino acid position",
       y = "Miyata score",
       title = "Miyata scores of fixed differences",
       subtitle = subtitle_label) +
    scale_x_continuous(limits = c(0, ncol(AA_aligned_df_NoNULL)))

} else if (exists("rect_df") == FALSE && exists("hypomorph") == TRUE) {
  print("Hypomorph SNP present to be annotated")
   final_plot = Hypomorph_SNP +
  geom_point(data = graph_data_divergences_only, size = 3, shape = 2) +
  aes(x = aminoacid_pos,
      y = Miyata_vector) +
  labs(x = "Amino acid position",
       y = "Miyata score",
       title = "Miyata scores of fixed differences",
       subtitle = subtitle_label) +
    scale_x_continuous(limits = c(0, ncol(AA_aligned_df_NoNULL)))

} else {
  print("Functional regions and hypomorph SNP present to be annotated")
  temp_plot = FxnlRegions + 
            geom_vline(data = hypomorph, 
            aes(xintercept = loc), 
              color = "purple4") 

  final_plot = temp_plot + 
  geom_point(data = graph_data_divergences_only, size = 3, shape = 2) +
  aes(x = aminoacid_pos,
      y = Miyata_vector) +
  labs(x = "Amino acid position",
       y = "Miyata score",
       title = "Miyata scores of fixed differences",
       subtitle = subtitle_label) +
    scale_x_continuous(limits = c(0, ncol(AA_aligned_df_NoNULL)))
}

final_plot
```