---
title: "Analyzing fixed differences between species"
author: "Miwa Wenzel"
date: 5/2/2020
output: github_document
---

# Set up
## Input files
### ***CHANGE THESE IF YOU WISH***
#### Input fasta files of aligned coding sequences of a gene and their respective names
##### Generally you would want the order of the sequences that you input to be in a phylogenetically reasonable order.
##### ***If you have a set of sequences that you want to evaluate against itself, set REFLEXIVE = TRUE.*** E.g. get Miyata scores for mel-sim, mel-yak, sim-yak. ***Alternatively, you may want to individually compare mel, sim, and yak to another set of sequences, such as predicted ancestral sequences. If this is the case, set REFLEXIVE = FALSE.*** Then you will have comparisons of mel-ancestral1, sim-ancestral1, yak-ancestral1, mel-ancestral2, sim-ancestral2, yak-ancestral2, mel-ancestral3, sim-ancestral3, yak-ancestral3. 
```{r input_catA}

REFLEXIVE = FALSE # ***CHANGE*** according to the above examples

nseqA = 8 # ***CHANGE*** to the number of input sequences of category A that you have

seqA1 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dsec_aligned.fasta" # ***CHANGE*** to your input file
seqA2 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dsim_aligned.fasta" # ***CHANGE*** to your input file
seqA3 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dmel_aligned.fasta" # ***CHANGE*** to your input file
seqA4 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dere_aligned.fasta" # ***CHANGE*** to your input file
seqA5 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dtei_aligned.fasta" # ***CHANGE*** to your input file
seqA6 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dyak_aligned.fasta" # ***CHANGE*** to your input file
seqA7 = "~/github/Sequence_Analysis/input_data/Miyata_score/Deug_aligned.fasta" # ***CHANGE*** to your input file
seqA8 = "~/github/Sequence_Analysis/input_data/Miyata_score/Dpse_aligned.fasta" # ***CHANGE*** to your input file

seqA_names = c("D.sec", "D.sim", "D.mel", "D.ere", "D.tei", "D.yak", "D.eug", "D.pse") # ***CHANGE*** to reflect the names of your input sequences. The order should match the order that you listed the sequences above.

seqA_label = "extant species" # ***CHANGE*** to reflect what you want your plot axis label to say
```

#### Change the below input if you have set REFLEXIVE = FALSE. Otherwise, ignore this section.
##### If you have REFLEXIVE = FALSE, then you will have a full heatmap grid. This may or may not be desirable for you. If it is not desirable for you, as in our case in which we are comparing extant species to ancestral nodes wherein not every extant species pairs with all the ancestral nodes, you can hide values from certain sequence pairs. See setting in the setup code below to do so.
```{r input_catB}

########### THE BELOW IS IMPORTANT IF 'REFLEXIVE = FALSE' BECAUSE YOU WANT TO COMPARE TWO SETS OF SEQUENCES. ##############

nseqB = 6 # ***CHANGE*** to the number of input sequences of categroy B that you have. If you are only dealing with one category, i.e. REFLEXIVE = TRUE, then this part (and anything subsequently with a "B" label) will be ignored.

seqB1 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node10_aligned.fasta" # ***CHANGE*** to your input file
seqB2 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node11_aligned.fasta" # ***CHANGE*** to your input file
seqB3 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node12_aligned.fasta" # ***CHANGE*** to your input file
seqB4 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node13_aligned.fasta" # ***CHANGE*** to your input file
seqB5 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node14_aligned.fasta" # ***CHANGE*** to your input file
seqB6 = "~/github/Sequence_Analysis/input_data/Miyata_score/Node15_aligned.fasta" # ***CHANGE*** to your input file

seqB_names = c("Node10", "Node11", "Node12", "Node13", "Node14", "Node15") # ***CHANGE*** to reflect the names of your input sequences. The order should match the order that you listed the sequences above.

seqB_label = "ancestral nodes" # ***CHANGE*** to reflect what you want your plot axis label to say for this input category

# You will now indicate whether you want the entire heatmap displayed and if not, which sequence pairs to grey out.
PLOT_ALL = FALSE # ***CHANGE*** Set to TRUE if you want a heatmap with every sequence pair's value displayed; set to FALSE if you want to omit the values of some pairs. If set to FALSE, indicate which pairs you want to omit in the vector below; make sure the names are exactly as written in the seqA_names and seqB_names vectors and to leave a single space between the names. 

plot_omit = c("D.pse Node10", "D.pse Node11", "D.pse Node12", "D.pse Node13", "D.pse Node14", "D.pse Node15",
              "D.eug Node11", "D.eug Node12", "D.eug Node13", "D.eug Node14", "D.eug Node15",
              "D.ere Node12", "D.ere Node13",
              "D.tei Node12", "D.tei Node13",
              "D.yak Node12", "D.yak Node13",
              "D.mel Node14", "D.mel Node15",
              "D.sec Node14", "D.sec Node15",
              "D.sim Node14", "D.sim Node15")

```

#### Labels and plot size input information
```{r labels}

subtitle_label = "bam" # ***CHANGE*** to the name of your gene or any other descriptor that you want in the output heatmap
plotwidth = 8 # ***CHANGE*** to set the width of the output plot in inches
plotheight = 8 # ***CHANGE*** to set the width of the output plot in inches

```

## Functions
### ***DON'T CHANGE THESE OR ANYTHING ELSE BELOW... unless you wanna fix my code :)***
####Created one big function that has all my common smaller functions to set up the input sequences in usable formats and get the total Miyata scores. Hope this is kosher.
```{r nt_to_Miyata}

library(seqinr)
library(stringr)

nt_to_Miyata = function(fasta1, fasta2) {

  get_sequences = function(fasta) {
    sequence_list = read.fasta(fasta, seqtype = "DNA", seqonly = TRUE)
    sequence_list2 = gsub(" ", "", sequence_list)
    sequence_names = c()
  
    for (i in 1:length(sequence_list)) {
      sequence_name = {}
      sequence_name = paste("sequence", i, sep = "")
      sequence_names = c(sequence_names, sequence_name)
    }
  
    sequence = mapply(assign, sequence_names, sequence_list2)
    return(as.data.frame(sequence))
  }

  get_codons = function(Sequences) {
    Codons = data.frame(matrix(nrow = nrow(Sequences), ncol = str_length(Sequences[1,1])/3))

    for (b in 1:nrow(Sequences)) {
      codon_vector = c()
      for (i in seq(1, str_length(Sequences[b,1]), 3)) { #break up the sequences into codons
        seq_codon = substr(Sequences[b,1], i, i + 2)
        codon_vector = c(codon_vector, seq_codon)
      }
      Codons[b,] = codon_vector
    }
    codons = Codons
  }

  codon_to_aa <- function(dna) {
    switch(dna, 
        "ATA" = "I", "ATT" = "I", "ATC" = "I",
        "CTT" = "L", "CTC" = "L", "CTA" = "L", "CTG" = "L", "TTA" = "L", "TTG" = "L",
        "GTT" = "V", "GTC" = "V", "GTA" = "V", "GTG" = "V",
        "TTT" = "F", "TTC" = "F",
        "ATG" = "M",
        "TGT" = "C", "TGC" = "C",
        "GCT" = "A", "GCC" = "A", "GCA" = "A", "GCG" = "A",
        "GGT" = "G", "GGC" = "G", "GGA" = "G", "GGG" = "G",
        "CCT" = "P", "CCC" = "P", "CCA" = "P", "CCG" = "P",
        "ACT" = "T", "ACC" = "T", "ACA" = "T", "ACG" = "T",
        "TCT" = "S", "TCC" = "S", "TCA" = "S", "TCG" = "S", "AGT" = "S", "AGC" = "S",
        "TAT" = "Y", "TAC" = "Y",
        "TGG" = "W",
        "CAA" = "Q", "CAG" = "Q",
        "AAT" = "N", "AAC" = "N",
        "CAT" = "H", "CAC" = "H",
        "GAA" = "E", "GAG" = "E",
        "GAT" = "D", "GAC" = "D",
        "AAA" = "K", "AAG" = "K",
        "CGT" = "R", "CGC" = "R", "CGA" = "R", "CGG" = "R", "AGA" = "R", "AGG" = "R",
        "TAA" = "Stop", "TAG" = "Stop", "TGA" = "Stop", "---" = "NA",
        "C--" = "NA", "T--" = "NA", "G--" = "NA", "A--" = "NA", 
        "CC-" = "NA", "TT-" = "NA", "GG-" = "NA", "AA-" = "NA", 
        "--C" = "NA", "--T" = "NA", "--G" = "NA", "--A" = "NA", 
        "-CC" = "NA", "-TT" = "NA", "-GG" = "NA", "-AA" = "NA", 
        "CT-" = "NA", "TC-" = "NA", "-CT" = "NA", "-TC" = "NA", 
        "CA-" = "NA", "AC-" = "NA", "-CA" = "NA", "-AC" = "NA", 
        "CG-" = "NA", "GC-" = "NA", "-CG" = "NA", "-GC" = "NA", 
        "AT-" = "NA", "TA-" = "NA", "-AT" = "NA", "-TA" = "NA", 
        "AG-" = "NA", "GA-" = "NA", "-AG" = "NA", "-GA" = "NA", 
        "GT-" = "NA", "TG-" = "NA", "-GT" = "NA", "-TG" = "NA") 
  }

  get_AAs = function(Codon_df) {
    AAs = apply(Codon_df, c(1,2), codon_to_aa)
  }

##############

  assign_hash = Vectorize(assign, vectorize.args = c("x", "value"))
  get_hash = Vectorize(get, vectorize.args = c("x"))
  exists_hash = Vectorize(exists, vectorize.args = c("x"))

  hash = new.env(hash = TRUE, parent = emptyenv(), size = 418L)


  key = c("C P", "C A", "C G", "C S", "C T", "C Q", "C E", "C N", "C D", "C H", "C K", "C R", "C V", "C L", "C I", "C M", "C F", "C Y", "C W", 
        "P A", "P G", "P S", "P T", "P Q", "P E", "P N", "P D", "P H", "P K", "P R", "P V", "P L", "P I", "P M", "P F", "P Y", "P W", 
        "A G", "A S", "A T", "A Q", "A E", "A N", "A D", "A H", "A K", "A R", "A V", "A L", "A I", "A M", "A F", "A Y", "A W",
        "G S", "G T", "G Q", "G E", "G N", "G D", "G H", "G K", "G R", "G V", "G L", "G I", "G M", "G F", "G Y", "G W", 
        "S T", "S Q", "S E", "S N", "S D", "S H", "S K", "S R", "S V", "S L", "S I", "S M", "S F", "S Y", "S W", 
        "T Q", "T E", "T N", "T D", "T H", "T K", "T R", "T V", "T L", "T I", "T M", "T F", "T Y", "T W",
        "Q E", "Q N", "Q D", "Q H", "Q K", "Q R", "Q V", "Q L", "Q I", "Q M", "Q F", "Q Y", "Q W",
        "E N", "E D", "E H", "E K", "E R", "E V", "E L", "E I", "E M", "E F", "E Y", "E W",
        "N D", "N H", "N K", "N R", "N V", "N L", "N I", "N M", "N F", "N Y", "N W",
        "D H", "D K", "D R", "D V", "D L", "D I", "D M", "D F", "D Y", "D W",
        "H K", "H R", "H V", "H L", "H I", "H M", "H F", "H Y", "H W",
        "K R", "K V", "K L", "K I", "K M", "K F", "K Y", "K W",
        "R V", "R L", "R I", "R M", "R F", "R Y", "R W",
        "V L", "V I", "V M", "V F", "V Y", "V W",
        "L I", "L M", "L F", "L Y", "L W",
        "I M", "I F", "I Y", "I W",
        "M F", "M Y", "M W",
        "F Y", "F W",
        "Y W",
        
        "C NA", "P NA", "A NA", "G NA", "S NA", "T NA", "Q NA", "E NA", "N NA", "D NA", "H NA", "K NA", "R NA", "V NA", "L NA", "I NA", "M NA", "F NA", "Y NA",
  
        "P C", "A C", "G C", "S C", "T C", "Q C", "E C", "N C", "D C", "H C", "K C", "R C", "V C", "L C", "I C", "M C", "F C", "Y C", "W C", 
        "A P", "G P", "S P", "T P", "Q P", "E P", "N P", "D P", "H P", "K P", "R P", "V P", "L P", "I P", "M P", "F P", "Y P", "W P", 
        "G A", "S A", "T A", "Q A", "E A", "N A", "D A", "H A", "K A", "R A", "V A", "L A", "I A", "M A", "F A", "Y A", "W A",
        "S G", "T G", "Q G", "E G", "N G", "D G", "H G", "K G", "R G", "V G", "L G", "I G", "M G", "F G", "Y G", "W G", 
        "T S", "Q S", "E S", "N S", "D S", "H S", "K S", "R S", "V S", "L S", "I S", "M S", "F S", "Y S", "W S", 
        "Q T", "E T", "N T", "D T", "H T", "K T", "R T", "V T", "L T", "I T", "M T", "F T", "Y T", "W T",
        "E Q", "N Q", "D Q", "H Q", "K Q", "R Q", "V Q", "L Q", "I Q", "M Q", "F Q", "Y Q", "W Q",
        "N E", "D E", "H E", "K E", "R E", "V E", "L E", "I E", "M E", "F E", "Y E", "W E",
        "D N", "H N", "K N", "R N", "V N", "L N", "I N", "M N", "F N", "Y N", "W N",
        "H D", "K D", "R D", "V D", "L D", "I D", "M D", "F D", "Y D", "W D",
        "K H", "R H", "V H", "L H", "I H", "M H", "F H", "Y H", "W H",
        "R K", "V K", "L K", "I K", "M K", "F K", "Y K", "W K",
        "V R", "L R", "I R", "M R", "F R", "Y R", "W R",
        "L V", "I V", "M V", "F V", "Y V", "W V",
        "I L", "M L", "F L", "Y L", "W L",
        "M I", "F I", "Y I", "W I",
        "F M", "Y M", "W M",
        "Y F", "W F",
        "W Y",
        
        
        "NA C", "NA P", "NA A", "NA G", "NA S", "NA T", "NA Q", "NA E", "NA N", "NA D", "NA H", "NA K", "NA R", "NA V", "NA L", "NA I", "NA M", "NA F", "NA Y"
        )

  value = c(1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,
          
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          
          1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,
          
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
          
          )

  assign_hash(key, value, hash)

###############

  get_Miyata = function(aminoacidsA, aminoacidsB) {
  
    AA_aligned = rbind(aminoacidsA, aminoacidsB)
  
    Miyata_sum = 0
    Miyata_vector = c()

    for (i in 1:ncol(AA_aligned)) {
  
      if (AAsA[1, i] == AAsB[1, i]) {
#       print("Amino acids are conserved")
      } else {
          pairedAAs = paste(AA_aligned[1, i], AA_aligned[2,i])
#         print(pairedAAs)
    
          Miyata = hash[[pairedAAs]]
#         print(Miyata)
          Miyata_vector = c(Miyata_vector, Miyata)
#         Miyata_sum = Miyata_sum + Miyata
      }
    }
#   print(Miyata_vector)
    vectorsum = sum(Miyata_vector)
#   print(vectorsum)
#   print(Miyata_sum)
#   cat("The total Miyata score is", vectorsum)

    return(as.numeric(vectorsum)) 
  }


  SequencesA = get_sequences(fasta1)
  SequencesB = get_sequences(fasta2)

  CodonsA = get_codons(SequencesA)
  CodonsB = get_codons(SequencesB)

  AAsA = get_AAs(CodonsA)
  AAsB = get_AAs(CodonsB)

  Total_Miyata = get_Miyata(AAsA, AAsB)
  return(Total_Miyata)

}
```

## Get the total Miyata scores for each sequence pair
##### I set warnings to FALSE because I had annoying messages from the formating of the input fasta files that I was using.
```{r get_Miyata, warning=FALSE}

#Get a vector of all the sequences from the input code chunk
input_seq_vector = c()

if (REFLEXIVE == TRUE) {
  print("Working with one set of sequences")
  
  for (i in 1:nseqA) {
    temp = eval(as.symbol(paste("seqA", i , sep = "")))
    assign(seqA_names[i], temp)
  } 
  
  all_pairs_df = t(as.data.frame(combn(seqA_names, 2))) #Get all the possible sequence pairing combinations to calculate Miyata scores of
  
   
} else {
  print("Working with two sets of sequences")
  
  for (i in 1:nseqA) {
    tempA = eval(as.symbol(paste("seqA", i, sep = "")))
    assign(seqA_names[i], tempA)
  }  
  
  for (j in 1:nseqB) {
    tempB = eval(as.symbol(paste("seqB", j, sep = "")))
    assign(seqB_names[j], tempB)
  }
  
  all_pairs_df = as.data.frame(expand.grid(seqA_names, seqB_names))
  
}

score_vector = c()

for (j in 1:nrow(all_pairs_df)) {
  pair_name = paste(all_pairs_df[j,1], "-", all_pairs_df[j,2], sep = "")
  
  pair1 = eval(as.symbol(as.character(all_pairs_df[j,1])))
  pair2 = eval(as.symbol(as.character(all_pairs_df[j,2])))
  
  score = nt_to_Miyata(pair1, pair2)
  score_vector = c(score_vector, score)
  assign(pair_name, score)
  
  cat("The total Miyata score for the", pair_name, "pair is", score, "\n")
  
}


all_pairs_df2 = cbind(all_pairs_df, score_vector)
all_pairs_df2 = as.data.frame(all_pairs_df2)
colnames(all_pairs_df2)[colnames(all_pairs_df2) == "score_vector"] <- "Total_Miyata"

```

### Heat map of these scores
##### Not sure if it's desirable or not to have a bunch of NA values to grey out the parts of the heat map that are repetitive... I have done that so far here. But you can get rid of it by using data=all_pairs_df2 in the ggplot function if you wish; makes the heat map smaller. 
```{r plot1, fig.height = plotheight, fig.width = plotwidth}

library(ggplot2)

if (REFLEXIVE == TRUE) {
  
  all_pairs_df3 = all_pairs_df2
  all_pairs_df3$Total_Miyata = c(rep(NA, each=nrow(all_pairs_df2)))
  all_pairs_df3$V1 = all_pairs_df2$V2
  all_pairs_df3$V2 = all_pairs_df2$V1

  all_pairs_df4 = rbind(all_pairs_df2, all_pairs_df3)

  all_pairs_df5 = data.frame(seqA_names, seqA_names, c(rep(NA, each=length(seqA_names))))
  colnames(all_pairs_df5) = c("V1", "V2", "Total_Miyata")

  all_pairs_df6 = rbind(all_pairs_df4, all_pairs_df5)
  
  plot1 = ggplot(data=all_pairs_df6, aes(x = factor(V1, level = seqA_names), y = factor(V2, level = seqA_names))) +
            geom_tile(aes(fill = as.numeric(as.character(Total_Miyata)))) +
            geom_text(aes(label = Total_Miyata)) +
            scale_fill_gradient2(low = "aquamarine1", high = "aquamarine4") +
            labs(x = seqA_label, y = seqA_label, title = "Total Miyata scores between sequences", subtitle = subtitle_label, fill = "Miyata score")

} else {
  
  all_pairs_df3 = all_pairs_df2
  
  if (PLOT_ALL == TRUE) {
    
  plot1 = ggplot(data = all_pairs_df3, aes(y = factor(Var1, level = seqA_names), x = factor(Var2, level = seqB_names))) +
            geom_tile(aes(fill = as.numeric(as.character(Total_Miyata)))) +
            geom_text(aes(label = Total_Miyata)) +
            scale_fill_gradient2(low = "aquamarine1", high = "aquamarine4") +
            labs(x = seqB_label, y = seqA_label, title = "Total Miyata scores between sequences", subtitle = subtitle_label, fill = "Miyata score")  
  } else {
    
    for (k in 1:nrow(all_pairs_df2)) {
      
      temp_pair = paste(all_pairs_df2[k,1], all_pairs_df2[k,2], sep = " ")
      
      if (temp_pair %in% plot_omit) {
        
        all_pairs_df3[k,3] = NA
        
      }
      
    }
    
     plot1 = ggplot(data = all_pairs_df3, aes(y = factor(Var1, level = seqA_names), x = factor(Var2, level = seqB_names))) +
            geom_tile(aes(fill = as.numeric(as.character(Total_Miyata)))) +
            geom_text(aes(label = Total_Miyata)) +
            scale_fill_gradient2(low = "aquamarine1", high = "aquamarine4") +
            labs(x = seqB_label, y = seqA_label, title = "Total Miyata scores between sequences", subtitle = subtitle_label, fill = "Miyata score")  
    
  }
  
}

plot1

```

# General purpose Miyata information
## Below is some useful visuals on the distribution of all possible Miyata scores.

### A distibution of Miyata scores
```{r, fig.height=5, fig.width=5}

Miyata_matrix_v = c(1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06)

ggplot() +
  aes(Miyata_matrix_v) +
  geom_histogram(binwidth = 0.25) +
  labs(title = "Distribution of all possible Miyata scores", x = "Miyata score") +
  scale_x_continuous(breaks = seq(0, 5.25, by = 1))

summary(Miyata_matrix_v)

```

### Distribution of Miyata scores from each amino acid to the others
```{r, fig.height = 10, fig.width = 15}

cys = c(1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34)
pro = c(0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 1.33)
ala = c(0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 0.06, 1.39)
gly = c(0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 0.91, 0.97, 2.22)
ser = c(0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 0.85, 0.51, 0.56, 2.84)
thr = c(1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5, 0.89, 1.7, 0.9, 0.87, 1.45)
gln = c(0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42, 1.65, 1.12, 2.48, 1.92, 1.92, 2.48)
glu = c(0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08, 0.84, 1.83, 2.06, 2.78, 2.46, 2.48, 3.26)
asn = c(0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 0.85, 0.99, 1.4, 1.31, 1.96, 1.78, 1.8, 2.83)
asp = c(1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88, 0.65, 0.9, 1.47, 2.05, 1.87, 2.37, 2.37, 2.4, 3.48)
his = c(0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16, 1.72, 1.29, 0.96, 0.32, 1.32, 1.94, 2.78, 2.17, 2.15, 2.56)
lys = c(0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11, 0.79, 2.05, 1.84, 1.14, 1.06, 2.1, 2.71, 3.54, 2.96, 2.94, 3.27)
arg = c(2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 0.4, 0.82, 2.34, 2.04, 1.45, 1.13, 2.03, 2.74, 3.58, 2.92, 2.9, 3.06)
val = c(0.91, 0.85, 0.62, 1.43, 1.52, 2.51, 2.43, 2.7, 2.11, 3.4, 2.76, 2.97, 2.13, 1.42, 2.15, 2.76, 1.85, 1.79, 0.86)
leu = c(0.14, 0.41, 0.63, 0.94, 1.73, 0.91, 2.62, 2.98, 2.59, 4.1, 3.49, 3.53, 2.7, 2.25, 3.04, 3.67, 2.276, 2.7, 1.65)
ile = c(0.29, 0.61, 0.86, 1.72, 0.14, 0.85, 2.49, 2.84, 2.45, 3.98, 3.37, 3.39, 2.57, 2.14, 2.95, 3.6, 2.69, 2.62, 1.63)
met = c(0.82, 0.93, 1.89, 0.29, 0.41, 0.92, 2.29, 2.63, 2.19, 3.69, 3.08, 3.13, 2.3, 1.86, 2.67, 3.34, 2.42, 2.36, 1.46)
phe = c(0.48, 1.11, 0.82, 0.61, 0.63, 1.43, 2.47, 2.85, 2.63, 4.27, 3.7, 3.59, 2.81, 2.6, 3.45, 4.14, 3.23, 3.17, 2.24)
tyr = c(1.06, 0.48, 0.93, 0.86, 0.94, 1.52, 2.02, 2.42, 2.27, 3.95, 3.42, 3.22, 2.48, 2.45, 3.33, 4.08, 3.18, 3.12, 2.38)
trp = c(1.06, 1.11, 1.89, 1.72, 1.73, 2.51, 2.72, 3.11, 3.16, 4.88, 4.39, 4.08, 3.42, 3.5, 4.38, 5.13, 4.23, 4.17, 3.34)

Miyata_df = data.frame(cys, pro, ala, gly, ser, thr, gln, glu, asn, asp, his, lys, arg, val, leu, ile, met, phe, tyr, trp)

aa_names = c("cys", "pro", "ala", "gly", "ser", "thr", "gln", "glu", "asn", "asp", "his", "lys", "arg", "val", "leu", "ile", "met", "phe", "tyr", "trp")

library(grid)

plots = list()
stats_names = c("Min", "1st Q", "Median", "Mean", "3rd Q", "Max")


mins = c()
firstqs = c()
medians = c()
means = c()
thirdqs = c()
maxs = c()


for (i in names(Miyata_df)) {
  stats = as.character(summary(Miyata_df[,i]))
  text1 = paste(stats_names[1], stats[1], "|", stats_names[2], stats[2], "|", stats_names[3], stats[3])
  text2 = paste(stats_names[4], stats[4], "|", stats_names[5], stats[5], "|", stats_names[6], stats[6])
  
#  grob = grobTree(textGrob(text1, x = 0.1, y = 0.95, hjust = 0, gp=gpar(col = "red", fontsize = 10, fontface = "italic")))
#  grob2 = grobTree(textGrob(text2, x = 0.1, y = 0.85, hjust = 0, gp=gpar(col = "red", fontsize = 10, fontface = "italic")))
  
    plots[[i]] = ggplot(data = Miyata_df) + 
    geom_histogram(aes_string(x=i), binwidth = 0.25) +
      scale_x_continuous(limits = c(0, 5)) +
      scale_y_continuous(limits = c(0,5)) +
    labs(title = i, x = "Miyata score")
#      annotation_custom(grob) +
#      annotation_custom(grob2)
    
    mintemp = stats[1]
    firstqtemp = stats[2]
    mediantemp = stats[3]
    meantemp = stats[4] 
    thirdqtemp = stats[5]
    maxtemp = stats[6]
  
  mins = c(mins, mintemp)
  firstqs = c(firstqtemp, firstqs)
  medians = c(medians, mediantemp)
  means = c(means,meantemp)
  thirdqs = c(thirdqs, thirdqtemp)
  maxs = c(maxs, maxtemp)
    
}

stats_df = data.frame(aa_names, mins, firstqs, medians, means, thirdqs, maxs)

library(gridExtra)
allplots = marrangeGrob(plots[1:20], nrow = 4, ncol = 5)


halfplots2 = marrangeGrob(plots[10:20], nrow = 5, ncol = 2)

allplots
print(stats_df)

```
