---
title: "Analyzing polymorphisms a population"
author: "Miwa Wenzel"
date: "5/1/2020"
output: github_document

---

# Set up
## Input files

### ***CHANGE THESE IF YOU WISH***
#### Input a multifasta file of the aligned sequences for a gene from a population.
```{r input}

input_fasta = "~/github/Sequence_Analysis/input_data/mel-bam-ZI-CDS-NoN.fasta" # ***CHANGE*** to the multifasta file of your aligned population data
```

#### Set up labels and stuff for the output graphs 
```{r labels}

subtitle_label = "bam | D. mel | Zambia | n=187" # ***CHANGE*** to reflect your gene and population information; this will go on the output graphs

########## SET FIGURE SIZES ##########

histwidth = 5 # ***CHANGE*** to set the width of the polymorphism output histogram in inches
histheight = 5 # ***CHANGE*** to set the height of the polymorphism output histogram in inches

plotwidth = 14 # ***CHANGE*** to set the width of the polymorphism output plot in inches
plotheight = 3 # ***CHANGE*** to set the height of the polymorphism output plot in inches


########## SET UP REGIONS TO ANNOTATE ON THE PLOT ##########

#This is a dataframe of important regions to highlight; idk why, but ordering the regions in this way makes the legend be in the desired order. Positions are in nucleotides, with the first nucleotide in the alignment being at position 1.

# ***CHANGE*** the notations as you wish OR if you don't want to highlight anything, comment these out!!

rect_df = data.frame(xmin = c(36, 451, 1204, 601), # ***CHANGE*** if want to highlight different regions for your gene
                     xmax = c(99, 1050, 1302, 750), # ***CHANGE*** to correspond to ends of the regions that you put in the above "xmin" section (e.g. the first region is from basepair position 451 to 1050)
                     ymin = c(-Inf, -Inf, -Inf, -Inf), 
                     ymax = c(Inf, Inf, Inf, Inf),
                     Region = factor(c("CBM", "Bgcn/Csn4 binding", "PEST domain", "Ubiquitin interaction"))) # ***CHANGE*** to reflect the names of your regions

rect_colors = c("#CC79A7", "steelblue1", "#009E73", "#F0E442") # ***CHANGE*** if you want to use different colors in highlighting your regions

hypomorph = data.frame(loc = 763, hypomorph = "L255F") # ***CHANGE*** if you want to highlight any position with a vertical line, e.g. like it currently points to the hypomorph position; I can't get this to show up in the legend however. 

```

## Functions
#### ***DON'T CHANGE THESE OR ANYTHING ELSE BELOW... unless you wanna fix my code :)***
##### This is a set of functions to get sequence data into a few different usable forms. Not all may be relevant for this code, but I maintain them together for ease. 
```{r sequence_functions}

############# IMPORT FASTA FILES AND CREATE DATAFRAME OF THE SEQUENCES ###################

library(seqinr) #load seqinr library in order to use the read.fasta function below

get_sequences = function(fasta) { #import sequences from fasta format
  sequence_list = read.fasta(fasta, seqtype = "DNA", seqonly = TRUE)
  sequence_list2 = gsub(" ", "", sequence_list) #remove any spaces in the fasta file if they are present to make sequences compatible with downstream functions
  sequence_names = c() #set empty vector to create sequence names to identify each sequence in the fasta file
  
  for (i in 1:length(sequence_list)) { #for each sequence in the fasta file, give it an identity of "sequence1, sequence2, etc." Then compile these such that we have a dataframe where the first column is the sequence names and the second column is the corresponding sequence itself)
    sequence_name = {}
    sequence_name = paste("sequence", i, sep = "")
    sequence_names = c(sequence_names, sequence_name)
  }
  sequence = mapply(assign, sequence_names, sequence_list2)
  as.data.frame(sequence)
}

############# CREATE A DATAFRAME OF NUCLEOTIDES FOR THE SEQUENCES ##################

library(stringr) #load library to use str_split_fixed function below

get_nucleotides = function(Sequences) { #Create a dataframe such that each nucleotide is in its own column. Thus we get a very large matrix where the rows are the different sequences and the columns are the nucleotides in the sequences.
  nucleotides = str_split_fixed(Sequences$sequence, "", str_length(Sequences[1,1]))
}

############# CREATE A DATAFRAME OF CODONS FOR THE SEQUENCES ##############

get_codons = function(Sequences) { #Now instead of splitting up the sequences into individual nucleotides for a dataframe, we want to split it up into codons
  Codons = data.frame(matrix(nrow = nrow(Sequences), ncol = str_length(Sequences[1,1])/3))

  for (b in 1:nrow(Sequences)) {
    codon_vector = c()
    for (i in seq(1, str_length(Sequences[b,1]), 3)) { #break up the sequences into codons
      seq_codon = substr(Sequences[b,1], i, i + 2)
      codon_vector = c(codon_vector, seq_codon)
    }
    Codons[b,] = codon_vector
  }
  codons = Codons
}

############# CREATE A DATAFRAME OF AMINO ACIDS FOR THE SEQUENCES #############

codon_to_aa = function(dna) {
 switch(dna, 
        "ATA" = "I", "ATT" = "I", "ATC" = "I",
        "CTT" = "L", "CTC" = "L", "CTA" = "L", "CTG" = "L", "TTA" = "L", "TTG" = "L",
        "GTT" = "V", "GTC" = "V", "GTA" = "V", "GTG" = "V",
        "TTT" = "F", "TTC" = "F",
        "ATG" = "M",
        "TGT" = "C", "TGC" = "C",
        "GCT" = "A", "GCC" = "A", "GCA" = "A", "GCG" = "A",
        "GGT" = "G", "GGC" = "G", "GGA" = "G", "GGG" = "G",
        "CCT" = "P", "CCC" = "P", "CCA" = "P", "CCG" = "P",
        "ACT" = "T", "ACC" = "T", "ACA" = "T", "ACG" = "T",
        "TCT" = "S", "TCC" = "S", "TCA" = "S", "TCG" = "S", "AGT" = "S", "AGC" = "S",
        "TAT" = "Y", "TAC" = "Y",
        "TGG" = "W",
        "CAA" = "Q", "CAG" = "Q",
        "AAT" = "N", "AAC" = "N",
        "CAT" = "H", "CAC" = "H",
        "GAA" = "E", "GAG" = "E",
        "GAT" = "D", "GAC" = "D",
        "AAA" = "K", "AAG" = "K",
        "CGT" = "R", "CGC" = "R", "CGA" = "R", "CGG" = "R", "AGA" = "R", "AGG" = "R",
        "TAA" = "Stop", "TAG" = "Stop", "TGA" = "Stop", "NNN" = "NA",
        "CNN" = "NA", "TNN" = "NA", "GNN" = "NA", "ANN" = "NA", 
        "CCN" = "NA", "TTN" = "NA", "GGN" = "NA", "AAN" = "NA", 
        "NNC" = "NA", "NNT" = "NA", "NNG" = "NA", "NNA" = "NA", 
        "NCC" = "NA", "NTT" = "NA", "NGG" = "NA", "NAA" = "NA", 
        "CTN" = "NA", "TCN" = "NA", "NCT" = "NA", "NTC" = "NA", 
        "CAN" = "NA", "ACN" = "NA", "NCA" = "NA", "NAC" = "NA", 
        "CGN" = "NA", "GCN" = "NA", "NCG" = "NA", "NGC" = "NA", 
        "ATN" = "NA", "TAN" = "NA", "NAT" = "NA", "NTA" = "NA", 
        "AGN" = "NA", "GAN" = "NA", "NAG" = "NA", "NGA" = "NA", 
        "GTN" = "NA", "TGN" = "NA", "NGT" = "NA", "NTG" = "NA") 
}

get_AAs = function(Codon_df) {
  AAs = apply(Codon_df, c(1,2), codon_to_aa)
}

```

### This is the set up for getting the Miyata scores for the polymorphisms
##### Use of hash function taken from here: https://blog.ephorie.de/hash-me-if-you-can
```{r Miyata_function, results=FALSE}

assign_hash = Vectorize(assign, vectorize.args = c("x", "value"))
get_hash = Vectorize(get, vectorize.args = c("x"))
exists_hash = Vectorize(exists, vectorize.args = c("x"))

hash = new.env(hash = TRUE, parent = emptyenv(), size = 418L)


key = c("C P", "C A", "C G", "C S", "C T", "C Q", "C E", "C N", "C D", "C H", "C K", "C R", "C V", "C L", "C I", "C M", "C F", "C Y", "C W", 
        "P A", "P G", "P S", "P T", "P Q", "P E", "P N", "P D", "P H", "P K", "P R", "P V", "P L", "P I", "P M", "P F", "P Y", "P W", 
        "A G", "A S", "A T", "A Q", "A E", "A N", "A D", "A H", "A K", "A R", "A V", "A L", "A I", "A M", "A F", "A Y", "A W",
        "G S", "G T", "G Q", "G E", "G N", "G D", "G H", "G K", "G R", "G V", "G L", "G I", "G M", "G F", "G Y", "G W", 
        "S T", "S Q", "S E", "S N", "S D", "S H", "S K", "S R", "S V", "S L", "S I", "S M", "S F", "S Y", "S W", 
        "T Q", "T E", "T N", "T D", "T H", "T K", "T R", "T V", "T L", "T I", "T M", "T F", "T Y", "T W",
        "Q E", "Q N", "Q D", "Q H", "Q K", "Q R", "Q V", "Q L", "Q I", "Q M", "Q F", "Q Y", "Q W",
        "E N", "E D", "E H", "E K", "E R", "E V", "E L", "E I", "E M", "E F", "E Y", "E W",
        "N D", "N H", "N K", "N R", "N V", "N L", "N I", "N M", "N F", "N Y", "N W",
        "D H", "D K", "D R", "D V", "D L", "D I", "D M", "D F", "D Y", "D W",
        "H K", "H R", "H V", "H L", "H I", "H M", "H F", "H Y", "H W",
        "K R", "K V", "K L", "K I", "K M", "K F", "K Y", "K W",
        "R V", "R L", "R I", "R M", "R F", "R Y", "R W",
        "V L", "V I", "V M", "V F", "V Y", "V W",
        "L I", "L M", "L F", "L Y", "L W",
        "I M", "I F", "I Y", "I W",
        "M F", "M Y", "M W",
        "F Y", "F W",
        "Y W",
        
        "C NA", "P NA", "A NA", "G NA", "S NA", "T NA", "Q NA", "E NA", "N NA", "D NA", "H NA", "K NA", "R NA", "V NA", "L NA", "I NA", "M NA", "F NA", "Y NA",
  
        "P C", "A C", "G C", "S C", "T C", "Q C", "E C", "N C", "D C", "H C", "K C", "R C", "V C", "L C", "I C", "M C", "F C", "Y C", "W C", 
        "A P", "G P", "S P", "T P", "Q P", "E P", "N P", "D P", "H P", "K P", "R P", "V P", "L P", "I P", "M P", "F P", "Y P", "W P", 
        "G A", "S A", "T A", "Q A", "E A", "N A", "D A", "H A", "K A", "R A", "V A", "L A", "I A", "M A", "F A", "Y A", "W A",
        "S G", "T G", "Q G", "E G", "N G", "D G", "H G", "K G", "R G", "V G", "L G", "I G", "M G", "F G", "Y G", "W G", 
        "T S", "Q S", "E S", "N S", "D S", "H S", "K S", "R S", "V S", "L S", "I S", "M S", "F S", "Y S", "W S", 
        "Q T", "E T", "N T", "D T", "H T", "K T", "R T", "V T", "L T", "I T", "M T", "F T", "Y T", "W T",
        "E Q", "N Q", "D Q", "H Q", "K Q", "R Q", "V Q", "L Q", "I Q", "M Q", "F Q", "Y Q", "W Q",
        "N E", "D E", "H E", "K E", "R E", "V E", "L E", "I E", "M E", "F E", "Y E", "W E",
        "D N", "H N", "K N", "R N", "V N", "L N", "I N", "M N", "F N", "Y N", "W N",
        "H D", "K D", "R D", "V D", "L D", "I D", "M D", "F D", "Y D", "W D",
        "K H", "R H", "V H", "L H", "I H", "M H", "F H", "Y H", "W H",
        "R K", "V K", "L K", "I K", "M K", "F K", "Y K", "W K",
        "V R", "L R", "I R", "M R", "F R", "Y R", "W R",
        "L V", "I V", "M V", "F V", "Y V", "W V",
        "I L", "M L", "F L", "Y L", "W L",
        "M I", "F I", "Y I", "W I",
        "F M", "Y M", "W M",
        "Y F", "W F",
        "W Y",
        
        
        "NA C", "NA P", "NA A", "NA G", "NA S", "NA T", "NA Q", "NA E", "NA N", "NA D", "NA H", "NA K", "NA R", "NA V", "NA L", "NA I", "NA M", "NA F", "NA Y"
        )

value = c(1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,
          
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          
          1.33, 1.39, 2.22, 2.84, 1.45, 2.48, 3.26, 2.83, 3.48, 2.56, 3.27, 3.06, 0.86, 1.65, 1.63, 1.46, 2.24, 2.38, 3.34,
          0.06, 0.97, 0.56, 0.87, 1.92, 2.48, 1.8, 2.4, 2.15, 2.94, 2.9, 1.79, 2.7, 2.62, 2.36, 3.17, 3.12, 4.17, 
          0.91, 0.51, 0.9, 1.92, 2.46, 1.78, 2.37, 2.17, 2.96, 2.92, 1.85, 2.76, 2.69, 2.42, 3.23, 3.18, 4.23, 
          0.85, 1.7, 2.48, 2.78, 1.96, 2.37, 2.78, 3.54, 3.58, 2.76, 3.67, 3.6, 3.34, 4.14, 4.08, 5.13, 
          0.89, 1.65, 2.06, 1.31, 1.87, 1.94, 2.71, 2.74, 2.15, 3.04, 2.95, 2.67, 3.45, 3.33, 4.38, 
          1.12, 1.83, 1.4, 2.05, 1.32, 2.1, 2.03, 1.42, 2.25, 2.14, 1.86, 2.6, 2.45, 3.5,
          0.84, 0.99, 1.47, 0.32, 1.06, 1.13, 2.13, 2.7, 2.57, 2.3, 2.81, 2.48, 3.42,
          0.85, 0.9, 0.96, 1.14, 1.45, 2.97, 3.53, 3.39, 3.13, 3.59, 3.22, 4.08,
          0.65, 1.29, 1.84, 2.04, 2.76, 3.49, 3.37, 3.08, 3.7, 3.42, 4.39, 
          1.72, 2.05, 2.34, 3.4, 4.1, 3.98, 3.69, 4.27, 3.95, 4.88,
          0.79, 0.82, 2.11, 2.59, 2.45, 2.19, 2.63, 2.27, 3.16,
          0.4, 2.7, 2.98, 2.84, 2.63, 2.85, 2.42, 3.11,
          2.43, 2.62, 2.49, 2.29, 2.47, 2.02, 2.72, 
          0.91, 0.85, 0.62, 1.43, 1.52, 2.51,
          0.14, 0.41, 0.63, 0.94, 1.73, 
          0.29, 0.61, 0.86, 1.72,
          0.82, 0.93, 1.89,
          0.48, 1.11,
          1.06,
          
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
          
          )

assign_hash(key, value, hash)

```

### Run the above functions to get sequence data in usable format
```{r sequences}

Sequences = get_sequences(as.character(input_fasta)) #apply function to import population sequences from fasta file

Nucleotides = get_nucleotides(Sequences) #apply function to get the population sequences into a usable dataframe format

Codons = get_codons(Sequences) #apply function to get population sequences into codon format so we can translate to amino acids

AminoAcids = get_AAs(Codons) #apply function to get amino acids of the population sequences

```

# Do the polymorphism analysis
##### Change the "results = hide" in the initial r header to get verbose output that says things like, "Polymorphism at site 8 with derived allele freq 0.2)
```{r polymorphisms, results="hide"}

library(data.table)

syn_polymorphism = 0 #set up counting variables
nonsyn_polymorphism = 0
total_polymorphism = 0

polymorphism_vector = c() #set up vectors to hold polymorphism information
derived_allele_freq_vector = c()

ancestral_vector = c()
derived_vector = c()

for (i in 1:ncol(Nucleotides)) { #go through every nucleotide position in the population sequence alignment
  derived_allele = {}
  derived_allele_freq = {}
  derived_allele_freq_percent = {}
  
  if (length(unique(Nucleotides[,i])) == 1) { #if the site is conserved across all individuals in the population
    polymorphism_vector = c(polymorphism_vector, "X") #mark the site as conserved with an X
    derived_allele_freq_vector = c(derived_allele_freq_vector, 0) #indicate that the derived allele frequency is 0
    ancestral_vector = c(ancestral_vector, unique(Nucleotides[,i])) #add the nucleotide to the vector that keeps track of ancestral nucleotides
    derived_vector = c(derived_vector, NA) #add a placeholder NA to the vector that keeps track of the derived nucleotides
    
  }    
  
  else { #if there are different nucleotides present at this position in the alignment...
      cat("Nucleotide position:", i,"\n") #print the nucleotide position
      cat("Polymorphism:",(unique(Nucleotides[,i])), "\n") #and print what the polymorphisms are
      total_polymorphism = total_polymorphism + 1 #add a count for the polymorphisms found
      
      if ((i %% 3) == 0) { #We want to determine if the polymorphism is synonymous or nonsynonymous. Here, we first check if the nucleotide position is evenly divisible by three so we can properly relate it to its amino acid.
        cat("Codon/aa position:", (i/3), "\n") #print the codon/amino acid position that this corresponds to 
        cat("Codons with polymorphism:", unique(Codons[,(i/3)]), "\n") #and print what the codons are
        cat("Amino acids:", unlist(unique(AminoAcids[,(i/3)])), "\n") #and print what the translated amino acids are 
        
        if (length(unique(AminoAcids[,(i/3)])) == 1) { #if the sequences translate to the same amino acid at this position...
          syn_polymorphism = syn_polymorphism + 1 #add a count to the synonymous polymorphism tally
          cat("synonymous polymorphism \n") #and print that this is a synonymous polymorphism
          
          polymorphism_vector = c(polymorphism_vector, "S") #keep track that this is a synonymous polymorphism at this site
          
          a = unlist(AminoAcids[, (i/3)]) #a series of commands to set up a table to determine the derived allele and its frequency
          AA_freq_table = data.table(a)
          
          t = unlist(Nucleotides[, i])

          allele_freq_table = data.table(t)
          setkey(allele_freq_table, t)
          sorted = allele_freq_table[,.N,by=list(t)]

          derived_allele = sorted[order(N)]$t[1] #Get the derived allele from the aforementioned table
          ancestral_allele = sorted[order(N)]$t[2] #Get the ancestral allele the aforementioned table
          
          derived_vector = c(derived_vector, derived_allele) #keep track of the derived allele in a vector
          ancestral_vector = c(ancestral_vector, ancestral_allele)
          
          derived_allele_freq = sorted[order(N)]$N[1] #Get the derived allele's frequency (aka the number of occurrences)
          derived_allele_freq_percent = derived_allele_freq/nrow(Nucleotides) #Get the derived allele's frequency as a percent of the population
          
          derived_allele_freq_vector = c(derived_allele_freq_vector, derived_allele_freq_percent) #Keep track of the derived allele frequency as a vector
          
          cat("Derived allele:", derived_allele, "with a frequency of", derived_allele_freq_percent, "\n \n")
          
        } else { #if the site (which is divisible by 3) is not synonymous, add a count to the nonsynonymous polymorphism tally and tell us it is nonsynonymous. Code below is repetitive to what is above.
          nonsyn_polymorphism = nonsyn_polymorphism + 1
          cat("nonsynonymous polymorphism \n")
          
          polymorphism_vector = c(polymorphism_vector, "N")
          
          t = unlist(Nucleotides[, i])

          allele_freq_table = data.table(t)
          setkey(allele_freq_table, t)
          sorted = allele_freq_table[,.N,by=list(t)]

          derived_allele = sorted[order(N)]$t[1]
          ancestral_allele = sorted[order(N)]$t[2] 
          
          derived_vector = c(derived_vector, derived_allele)
          ancestral_vector = c(ancestral_vector, ancestral_allele)
          
          
          derived_allele_freq = sorted[order(N)]$N[1]
          derived_allele_freq_percent = derived_allele_freq/nrow(AminoAcids)
          
          derived_allele_freq_vector = c(derived_allele_freq_vector, derived_allele_freq_percent)
          
          cat("Derived allele:", derived_allele, "with a frequency of", derived_allele_freq_percent, "\n \n")
        }
        
      } else { #if the nucleotide position is not evenly divisible by three, then to get the right codon/amino acid position we have to take this into account and then I am repeating the code as was done above
        
        cat("Codon/aa position:", (i %/% 3 + 1), "\n")  
        cat("Codons with polymorphism:", unique(Codons[,(i/3) + 1]), "\n")
        cat("Amino acids:", unlist(unique(AminoAcids[,(i %/% 3 + 1)])), "\n")
        
        if (length(unique(AminoAcids[,(i %/% 3) + 1])) == 1) { #if nucleotides are identical at this site...
          syn_polymorphism = syn_polymorphism + 1 #add a count to the synonymous polymorphism counter
          cat("synonymous polymorphism \n") 
          
          polymorphism_vector = c(polymorphism_vector, "S")
          
          t = unlist(Nucleotides[, i])

          allele_freq_table = data.table(t)
          setkey(allele_freq_table, t)
          sorted = allele_freq_table[,.N,by=list(t)]

          derived_allele = sorted[order(N)]$t[1]
          ancestral_allele = sorted[order(N)]$t[2] #Get the ancestral allele
          
          derived_vector = c(derived_vector, derived_allele)
          ancestral_vector = c(ancestral_vector, ancestral_allele)
          
          derived_allele_freq = sorted[order(N)]$N[1]
          derived_allele_freq_percent = derived_allele_freq/nrow(Nucleotides)
          
          derived_allele_freq_vector = c(derived_allele_freq_vector, derived_allele_freq_percent)
          
          cat("Derived allele:", derived_allele, "with a frequency of", derived_allele_freq_percent, "\n \n")
          
        } else { #if nucleotides are not identical at this site...
          nonsyn_polymorphism = nonsyn_polymorphism + 1 #add a count to the nonsynonymous polymorphism counter
          cat("nonsynonymous polymorphism \n")
          
          polymorphism_vector = c(polymorphism_vector, "N")
          
          t = unlist(Nucleotides[, i]) #get all the nucleotides that are at this position

          allele_freq_table = data.table(t)
          setkey(allele_freq_table, t)
          sorted = allele_freq_table[,.N,by=list(t)] #Get a count of the ancestral and derived alleles

          derived_allele = sorted[order(N)]$t[1] #Get the derived allele
          ancestral_allele = sorted[order(N)]$t[2] #Get the ancestral allele
          
          derived_vector = c(derived_vector, derived_allele)
          ancestral_vector = c(ancestral_vector, ancestral_allele)
          
          derived_allele_freq = sorted[order(N)]$N[1]
          derived_allele_freq_percent = derived_allele_freq/nrow(Nucleotides) #get the frequency of the derived allele as a percent
          
          derived_allele_freq_vector = c(derived_allele_freq_vector, derived_allele_freq_percent) #store all the frequencies here
          
          cat("Derived allele:", derived_allele, "with a frequency of", derived_allele_freq_percent, "\n \n")
      }
    }
  } #end if statement to find polymorphisms in the population
} #end of for loop that cycles through all the nucleotides positions in the gene
```

## Do the Miyata analysis
```{r Miyata}

############### Get ancestral and derived amino acids ########

ancestral_AA_vector = c() #Set ourselves up to keep track of ancestral amino acids
derived_AA_vector = c() #Set ourselves up to keep track of derived amino acids

number_of_derived_vector = c() #Set ourselves up to keep track of the number of derived (aka non-synonymous) amino acids

for (i in 1:ncol(AminoAcids)) { #for every site in the translated (amino acid) alignment 
  
  if (length(unique(AminoAcids[,i])) == 1) { #if all the amino acids at this site are identical, then...

    derived_AA = NA #there are no derived alleles
    ancestral_AA = AminoAcids[1,i] #the ancestral allele is the allele that's present in the first (or any) sequence

    } else { #if there is variation in the amino acid that is present at this site, then...
    
    a = unlist(AminoAcids[, i]) #get all the amino acids that are at this position  
    AA_freq_table = data.table(a)
    setkey(AA_freq_table, a)
    sortedAA = AA_freq_table[,.N,by=list(a)] #create a table with the amino acids and their frequency counts
  
    derived_AA = sortedAA[order(N)]$a[1] #get the derived amino acid, which is the one that is less frequent
    ancestral_AA = sortedAA[order(N)]$a[2] #get the ancestral amino acid, which is the one that is more frequent
    
    number_of_derived = sortedAA[order(N)]$N[1] #get the number of instances that the derived (nonsynonymous) amino acid occurs
    number_of_derived_vector = c(number_of_derived_vector, number_of_derived)

    }
    
    derived_AA_vector = c(derived_AA_vector, derived_AA) #keep track of the derived amino acids
    ancestral_AA_vector = c(ancestral_AA_vector, ancestral_AA) #keep track of the ancestral amino acids
}

AA_polymorphisms = rbind(ancestral_AA_vector, derived_AA_vector) #create a dataframe that has for every site in the alignment the ancestral amino acid and, where they exist, the derived amino acid. (Note: these are not true, existing sequences.)

############### GET MIYATA SCORES FOR EACH NONSYNONYMOUS POLYMORPHISM ####################

Miyata_vector = c()

for (j in 1:ncol(AA_polymorphisms)) { #for every site in the aforementioned dataframe that has all the polymorphisms compiled...
  if (is.na(AA_polymorphisms[2,j]) == TRUE) { #if the amino acids are the same (no derived allele), then...
    Miyata_vector = c(Miyata_vector, 0) #give a Miyata score of 0
  } else { #if the amino acids are different, then...
      pairedAAs = paste(AA_polymorphisms[1, j], AA_polymorphisms[2,j]) #find out which two amino acids they are...

      Miyata = hash[[pairedAAs]] #...and get the Miyata score for those two amino acids
      Miyata_vector = c(Miyata_vector, Miyata) #keep track of these Miyata scores in a vector
  }
}

```

## Make dataframes to summarize our data and use in our graphs
#### First we have to adjust our amino acid data to match up with our nucleotide data.
```{r AminoAcid_to_Nuclotide_df}

rep.col<-function(x,n){ #a function to triple every element in the AA vectors so the values will line up with nucleotide vectors. Function taken from: https://www.r-bloggers.com/a-quick-way-to-do-row-repeat-and-col-repeat-rep-row-rep-col/
   matrix(rep(x,each=n), ncol=n, byrow=TRUE)
}

derived_AA_nt = rep.col(derived_AA_vector, 3) #apply the function to triple every value in our list of derived amino acids.
derived_AA_ntgraph = as.vector(t(derived_AA_nt)) #transpose the dataframe so it is in the proper format for us to use

ancestral_AA_nt = rep.col(ancestral_AA_vector, 3)
ancestral_AA_ntgraph = as.vector(t(ancestral_AA_nt))

Miyata_vector_nt = rep.col(Miyata_vector, 3)
Miyata_vector_ntgraph = as.vector(t(Miyata_vector_nt))

```

#### Next we can make the dataframes
```{r final_data}

nucleotide_position = seq(1, ncol(Nucleotides), by = 1)

graph_data2 = data.frame(nucleotide_position, ancestral_vector, derived_vector, derived_allele_freq_vector, polymorphism_vector, ancestral_AA_ntgraph, derived_AA_ntgraph, Miyata_vector_ntgraph) #dataframe that consists of the ancestral nucleotide and ancestral amino acid, and, if it occurs anywhere in the population at the given site, the derived nucleotide, derived amino acid, the derived allele frequency, the Miyata score between the derived allele and the ancestral allele, and an indication of sysnonymous/nonsynonymous.

graph_data_polymorphisms_only2 = graph_data2[!(graph_data2$polymorphism_vector=="X"),] #we remove all the "uninteresting" conserved sites to make a dataframe with just the polymorphisms.

```

### A simple analysis of the distribution of Miyata scores of the nonsynonymous polymorphisms
```{r histogram, fig.height = histheight, fig.width = histwidth}

############# ORGANIZE THE DATA #############
Miyata_vector_NoZero = Miyata_vector[!Miyata_vector %in% 0] #Get the Miyata scores for only sites that are nonsynonymous polymorphisms

Miyata_vector_all_polymorphism = c()

for (k in 1:length(Miyata_vector_NoZero)) { #for each Miyata score of the nonsynonymous polymorphisms...
  Miyata_rep = rep.col(Miyata_vector_NoZero[k], number_of_derived_vector[k]) #replicate it by the number of instances that the polymorphisms occurs in the population
  Miyata_vector_all_polymorphism = c(Miyata_vector_all_polymorphism, Miyata_rep) #store the Miyata scores in a vector that accurately shows how many times it occurs in the population
}

############# MAKE THE HISTOGRAMS #############

library(ggplot2)
library(dplyr)

#A histogram of all the Miyata scores, including lots of 0s for the conserved sites and synonymous polymorphisms
#Not that useful, so it is commented out currently
#ggplot() +
#  aes(Miyata_vector) +
#  geom_histogram(binwidth = 0.25) +
#  labs(title = "Miyata scores of population", subtitle = subtitle_label, x = "Miyata score") +
#  scale_x_continuous(breaks = seq(0, 5.25, by = 1))

#A histogram of the set of Miyata scores - no allele frequency is considered here - of the nonsynonymous polymorphisms only.
ggplot() +
  aes(Miyata_vector_NoZero) +
  geom_histogram(binwidth = 0.25) +
  labs(title = "Miyata scores of polymorphisms", subtitle = subtitle_label, x = "Miyata score") +
  scale_x_continuous(breaks = seq(0, 5.25, by = 1))

#A histogram of Miyata scores with allele frequency taken into account of nonsynonymous polymorphisms.
ggplot() +
  aes(Miyata_vector_all_polymorphism) +
  geom_histogram(binwidth = 0.25) +
  labs(title = "Miyata scores of polymorphisms with allele frequency", subtitle = subtitle_label, x = "Miyata score") +
  scale_x_continuous(breaks = seq(0, 5.25, by = 1))

```


## Make the final graph
##### Set up the background data. This is set to highlight documented functional domains of our gene bam. You can change this to reflect the known regions of your gene and/or not run it. **I HAVE MOVED THIS TO THE TOP OF THE DOCUMENT SO IT IS EASY TO ACCESS AND CHANGE. Thus this code chunk is defaulted to NOT be run.**
```{r graph_background, eval=FALSE}

#Below is data used for amino acid regions
#rect_df = data.frame(xmin = c(13, 151, 402, 201), #dataframe of important regions to highlight
#                     xmax = c(33, 350, 434, 250),
#                     ymin = c(-Inf, -Inf, -Inf, -Inf),
#                     ymax = c(Inf, Inf, Inf, Inf),
#                     Region = factor(c("bgcn/csn4 binding", "PEST domain", "Ubiquitin interaction")))

rect_df = data.frame(xmin = c(451, 36, 1204, 601), #dataframe of important regions to highlight; idk why, but order the regions in this way makes the legend be in the desired order
                     xmax = c(1050, 99, 1302, 750),
                     ymin = c(-Inf, -Inf, -Inf, -Inf),
                     ymax = c(Inf, Inf, Inf, Inf),
                     Region = factor(c("CBM", "Bgcn/Csn4 binding", "PEST domain", "Ubiquitin interaction")))


hypomorph = data.frame(loc = 763, hypomorph = "L255F") #dataframe of hypomorph position

```

#### Set up the different types of plots to prepare for whether or not we want to include annotations in our final plot 
```{r}

library(ggplot2)

Hypomorph_SNP = ggplot() +     
            geom_vline(data = hypomorph, 
            aes(xintercept = loc), 
              color = "purple4") 

FxnlRegions = ggplot() +
            geom_rect(data = rect_df,
            aes(xmin = xmin, 
                xmax = xmax, 
                ymin = ymin,
                ymax = ymax,
                x = NULL,
                y = NULL,
                fill = Region,
                color = NULL),
            alpha = 0.3) +
  
  scale_fill_manual(values = rect_colors)

Polymorphisms = ggplot() +
    geom_point(data = graph_data_polymorphisms_only2, size = 3) + 
  
  aes(x = nucleotide_position,
      y = derived_allele_freq_vector,
      color = as.numeric(Miyata_vector_ntgraph)) +
      scale_color_gradient(low = "white", high = "black") +
  
  labs(x = "Nucleotide position",
       y = "Derived allele frequency",
       title = "Polymorphism frequency and Miyata scores",
       subtitle = subtitle_label) +
  
    scale_x_continuous(limits = c(0, ncol(Nucleotides))) +
  
  geom_point(data = graph_data_polymorphisms_only2, shape = 1, size = 3, colour = "black") 


```

### The final plot of polymorphisms!
```{r final_graph, fig.height = plotheight, fig.width = plotwidth}

if (exists("rect_df") == FALSE && exists("hypomorph") == FALSE) {
  print("Regions and a hypomorph SNP are NOT present to be annotated")
  final_plot = Polymorphisms
  
} else if (exists("rect_df") == TRUE && exists("hypomorph") == FALSE) {
  print("Regions are present to be annotated")
  
  final_plot = FxnlRegions + 
    geom_point(data = graph_data_polymorphisms_only2, size = 3) + 
  
  aes(x = nucleotide_position,
      y = derived_allele_freq_vector,
      color = as.numeric(Miyata_vector_ntgraph)) +
      scale_color_gradient(low = "white", high = "black") +
  
  labs(x = "Nucleotide position",
       y = "Derived allele frequency",
       title = "Polymorphism frequency and Miyata scores",
       subtitle = subtitle_label) +
  
    scale_x_continuous(limits = c(0, ncol(Nucleotides))) +
  
  geom_point(data = graph_data_polymorphisms_only2, shape = 1, size = 3, colour = "black") 

} else if (exists("rect_df") == FALSE && exists("hypomorph") == TRUE) {
  print("Hypomorph SNP present to be annotated")
   final_plot = Hypomorph_SNP +
    geom_point(data = graph_data_polymorphisms_only2, size = 3) + 
  
  aes(x = nucleotide_position,
      y = derived_allele_freq_vector,
      color = as.numeric(Miyata_vector_ntgraph)) +
      scale_color_gradient(low = "white", high = "black") +
  
  labs(x = "Nucleotide position",
       y = "Derived allele frequency",
       title = "Polymorphism frequency and Miyata scores",
       subtitle = subtitle_label) +
  
    scale_x_continuous(limits = c(0, ncol(Nucleotides))) +
  
  geom_point(data = graph_data_polymorphisms_only2, shape = 1, size = 3, colour = "black") 

} else {
  print("Functional regions and hypomorph SNP present to be annotated")
  temp_plot = FxnlRegions + 
            geom_vline(data = hypomorph, 
            aes(xintercept = loc), 
              color = "purple4") 
  
  final_plot = temp_plot + 
        geom_point(data = graph_data_polymorphisms_only2, size = 3) + 
  
  aes(x = nucleotide_position,
      y = derived_allele_freq_vector,
      color = as.numeric(Miyata_vector_ntgraph)) +
      scale_color_gradient(low = "white", high = "black") +
  
  labs(x = "Nucleotide position",
       y = "Derived allele frequency",
       title = "Polymorphism frequency and Miyata scores",
       subtitle = subtitle_label) +
  
    scale_x_continuous(limits = c(0, ncol(Nucleotides))) +
  
  geom_point(data = graph_data_polymorphisms_only2, shape = 1, size = 3, colour = "black") 

}

final_plot

```
